File: 460-lfu-cache.py
from collections import defaultdict, OrderedDict

class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.key_to_val_freq = {}  # key -> (val, freq)
        self.freq_to_keys = defaultdict(OrderedDict)  # freq -> {key: None}
        self.min_freq = 0

    def get(self, key: int) -> int:
        if key not in self.key_to_val_freq:
            return -1

        val, freq = self.key_to_val_freq[key]

        # Remove key from current freq list
        del self.freq_to_keys[freq][key]

        # If that was the last key with min_freq, update min_freq
        if not self.freq_to_keys[freq]:
            if self.min_freq == freq:
                self.min_freq += 1
            del self.freq_to_keys[freq]

        # Increase frequency
        self.freq_to_keys[freq + 1][key] = None
        self.key_to_val_freq[key] = (val, freq + 1)

        return val

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return

        # If key exists, update its value and call get() to update freq
        if key in self.key_to_val_freq:
            self.key_to_val_freq[key] = (value, self.key_to_val_freq[key][1])  # update value
            self.get(key)  # also increases frequency
            return

        # Evict if at capacity
        if len(self.key_to_val_freq) >= self.capacity:
            # Remove the LRU key in min_freq
            evict_key, _ = self.freq_to_keys[self.min_freq].popitem(last=False)
            del self.key_to_val_freq[evict_key]

        # Insert new key
        self.freq_to_keys[1][key] = None
        self.key_to_val_freq[key] = (value, 1)
        self.min_freq = 1  # reset min_freq for new key

File: c++/460-lfu-cache.cpp
#include <unordered_map>
#include <list>
using namespace std;

class LFUCache {
private:
    int capacity;
    int minFreq;
    unordered_map<int, pair<int, int>> keyToValFreq;           // key -> {value, freq}
    unordered_map<int, list<int>> freqToKeys;                  // freq -> list of keys (LRU order)
    unordered_map<int, list<int>::iterator> keyToIter;         // key -> iterator in freqToKeys

public:
    LFUCache(int capacity) {
        this->capacity = capacity;
        minFreq = 0;
    }

    int get(int key) {
        if (keyToValFreq.find(key) == keyToValFreq.end())
            return -1;

        // Update frequency
        int val = keyToValFreq[key].first;
        int freq = keyToValFreq[key].second;
        keyToValFreq[key].second++;

        // Remove from old freq list
        freqToKeys[freq].erase(keyToIter[key]);

        // Clean up if freq list becomes empty and update minFreq
        if (freqToKeys[freq].empty()) {
            freqToKeys.erase(freq);
            if (minFreq == freq) minFreq++;
        }

        // Add to new freq list
        freqToKeys[freq + 1].push_front(key);
        keyToIter[key] = freqToKeys[freq + 1].begin();

        return val;
    }

    void put(int key, int value) {
        if (capacity == 0) return;

        if (keyToValFreq.find(key) != keyToValFreq.end()) {
            keyToValFreq[key].first = value;
            get(key);  // Update frequency
            return;
        }

        // Eviction
        if (keyToValFreq.size() >= capacity) {
            // Evict LRU from lowest freq
            int evictKey = freqToKeys[minFreq].back();
            freqToKeys[minFreq].pop_back();

            if (freqToKeys[minFreq].empty())
                freqToKeys.erase(minFreq);

            keyToValFreq.erase(evictKey);
            keyToIter.erase(evictKey);
        }

        // Insert new key
        keyToValFreq[key] = {value, 1};
        freqToKeys[1].push_front(key);
        keyToIter[key] = freqToKeys[1].begin();
        minFreq = 1;
    }
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
